design/namespaced-roles.md

```diff
@@ -0,0 +1,349 @@
+
+- Improve the experience of managing roles. By namespacing them, it would be easier to add custom structures that have business meaning for users.
+
+In order to have this flexibility and still support current use cases, we could create a new kind of role or backport existing roles to this new format: Namespaced roles.
+
+## Specification
+
+A Namespaced Role would contain a version, a path that start with a known particle to identify the nature of the role, and finally the role name.
+
+For example, Keycloak management specific roles would start with `/kc`.
+
+User defined roles, but still used to interact with Keycloak entities would start with `/ud`.
+
+Totally custom namespaces created with a specific business meaning for users wouldn't contain any of the previous particles.
+
+Namespaced Roles would always follow a structure that contains a fixed part identifying the entity, followed by a variable part containing its identifier. This pattern could be repeated in a hierarchical way.
+
+A few examples would be:
+
+`role_v1:/kc/groups/{group-name}/roleName`
+
+`role_v1:/ud/tenants/{tenant-name}/roleName`
+
+`role_v1:/ud/tenants/{tenant-name}/clients/{client-name}/roleName`
+
+Some of the Keycloak supported entities would be:
+
+**Tenant**: There's a proposal to make Keycloak multitenant aware, so even if it's not currently supported, for future compatibility, it should be considered when defining a unique identifier.
+   - Example: `role_v1:/kc/tenants/tenantA/roleName`.
+
+**Client**: Roles can be defined and assigned for a certain client in Keycloak, so a unique identifier has to support this possibility.
+   - Example: `role_v1:/kc/clients/clientId/roleName`.
+
+Alternatively, if a namespace wants to be created for roles that can be shared among clients, it could be defined such as:  `role_v1:/clients/*/roleName`.
+
+**Group**: Roles can be assigned to groups or, as intended in this design proposal, assigned to a user in the context of a group.
+   - Example: `role_v1:/groups/groupA/roleName`.
+
+Alternatively, if a namespace wants to be created for roles that can be shared among groups, it could be defined such as:  `role_v1:/groups/*/roleName`
+
+A free-form role identifier would be allowed for resource-server specific roles thay may hold business meaning for users. It would lack the fixed `kc` or `ud` particles and allow any user defined path. For example:
+
+- `role_v1:/business-units/myBusinessUnit/const-centers/myCostCenter/roleName`
+
+**Question**: In the initial proposal, these roles would be preceded by a fixed part that would identify the context where this role would be applied. We could do somethng similar here such as:
+
+- `role_v1:/groups/{group-name}/role-path/business-units/myBusinessUnit/const-centers/myCostCenter/roleName`
+
+But not convinced on the way to let Keycloak know that that's not a role name but rather a new namespace.
```

Comment:
I was thinking that clients, groups, tenants, etc, would instead of having a fixed path to their namespace would allow configuring the namespace that is associated with them.

So, a client could have it's default namespace be `/kc/clients/{client-id}`, but could be configured to use a different namespace like `/my-roles/my-group-of-apps`